### PR TITLE
WIP: Content addressed tarballs

### PR DESCRIPTION
**Pull request by scottwittenburg**
Store binary archives (compressed tarballs of spack install trees) by their computed checksum in order to remove the possibility of race conditions when multiple processes write binaries for the same spec hash simulatenously.  This PR is based on #40941 in hopes that can be merged and backported to the current release.  Otherwise, this PR can just include the commit from that one.

### Summary of the change

The change here is fairly small.  Essentially, tarballs will be stored in files named after their hash digest, and will live in a directory grouped with any other tarballs for the same spec hash.

<details>

<summary>Depiction of old and new layout</summary>

### Current buildcache layout (version 1)

```console
                                                                                                 tarball_path_name(spec, ext)
                                               |----------------------------------------------------------------------------------------------------------------------------------------|

s3://spack-binaries/releases/v0.20/build_cache/linux-ubuntu20.04-haswell/gcc-9.4.0/libiconv-1.17/linux-ubuntu20.04-haswell-gcc-9.4.0-libiconv-1.17-7kvjvhmmou4tljgebolmis2slnkgf34j.spack
|--------------------------------|             |-----------------------------------------------| |--------------------------------------------------------------------------------------|

         cache prefix                                      tarball_directory_name(spec)                               tarball_name(spec, ext) - gzip compressed tarball
```

### Proposed next buildcache layout (version 2)

```console
s3://spack-binaries/releases/v0.20/build_cache/linux-ubuntu20.04-haswell/gcc-9.4.0/libiconv-1.17/7kvjvhmmou4tljgebolmis2slnkgf34j/<sha256>
|--------------------------------|             |--------------------------------------------------------------------------------| |------|

         cache prefix                                                      relative_tarball_directory(spec)                         gzip compressed tarball
```
</details>

In both version 1 and 2 above, the spec metadata file will live in the same location and follow the
current naming rules, so for the above-mentioned `libiconv` spec, that is:

```console
s3://spack-binaries/releases/v0.20/build_cache/linux-ubuntu20.04-haswell-gcc-9.4.0-libiconv-1.17-7kvjvhmmou4tljgebolmis2slnkgf34j.spec.json[.sig]
```

The contents of the buildcache spec metadata under version 2 only differs from that of version 1 in two ways:

1. the "buildcache_layout_version" is now 2 instead of 1
2. an additional "digest" is included at the top level, e.g.: `"digest": "sha256:bd773ee67fbe4066c101014f324bee5b3d738d057f34711dab43a3ae8bf51222"`

The digest value can then be used to find the actual compressed tarball.

This PR has been implemented so that spack still supports reading the previous buildcache layouts (both version 1 and the initial, "unversioned" layout), in addition to version 2. A mirror is allowed to contain a mix of new and old formats.  This allows us to keep reusing binaries built/pushed to the mirrors before the change, while writing only the new format.

### Benefits of the new layout

- avoids race conditions we have seen in the past when multiple jobs are building binaries for the same spec hash simultaneously (e.g. gitlab fails to cancel the previous pipeline when new changes are pushed and it triggers the next pipeline)
- allows us to continue pruning mirrors based on spec hash (any compressed tarball named as <sha256 stored in tarball_directory_name(spec) can be deleted).
- takes a step towards an organization more like the new oci-buildcache feature

### Possible extensions of this effort

The approach outlined here will allow new spack to read old buildcache layouts, but when old spack encounters a binary using the newest layout, it will not find the compressed archive, and will be forced to install from source.  To help with this, this PR is based on #40941 which allows spack to notice and warn when it encounters a spec using a buildcache layout that is too new.  Perhaps that PR can be merged and backported into the current release, and then at least the previous release before the layout change is merged will support this warning behavior.  If this is not enough support for "old spack reading newer buildcache formats", a possible extension of this PR could include continuing to write the old layout during a "deprecation" period.  Some possible variants might be:

- write either the old or new layout, controlled by a cli option
- always write the new layout, but optionally write the old layout too (controlled by cli option)
- always write both layouts

Another possible extension could include support for automatic migration of buildcache layouts.  Spack could possibly provide a buildcache migration command to upgrade any specs within a mirror from version 1 to version 2 (something like `spack buildcache update-layout <mirror_name_or_url>`).  This is sort of the inverse of the situation described just above, where a newer spack intends to read the previous layout version.  This should be easy to implement as no changes are required to the stored archive, it only needs to be moved to a new location and renamed.  However, since this PR currently supports reading the previous layout in addition to the new one, it's maybe of questionable value to perform such a migration.

### CODE REVIEW

**Comment by scottwittenburg** [comment]
fyi @kwryankrattiger

**Comment by haampie** [comment]
I think this PR is a good opportunity to discuss binary vs source.

TL;DR don't invent a new naming scheme for each instance where something needs to be content addressed -- instead keep it generic so you can lookup content in your favorite directory with blobs.

What I don't like about the suggested scheme is that this keeps the distinction between source and binary downloads in Spack. That's not a useful distinction. If you make those the same thing, you get the benefits of download cache that exists in "source mirrors" but does not in "binary mirrors".

Taking that a step further, also there is no point in making a "download cache" a special concept in Spack, you can have download cache = build cache = source cache, it can all "fetch" the same way. (Download cache may be lacking metadata files, although they exist for source cache now).

The only minor difference with source / binary cache is that metadata lives in different places, shasums for source cache come from `package.py,` whereas for binaries the come from `spec.json`, so it may still be worth splitting the binaries into two locations to make pruning easier. But it can still reuse code. (Or have a layer of indirection through a common metadata file, like manifests in oci storage, then pruning can be generic for combined source/binary caches)

Also regarding naming scheme, follow either Spack's existing source cache structure, or a slight improvement on that, include the hash algorithm in the name too. OCI scheme just uses e.g. `sha256:<shasum>` as a file name (except that the colon is probably forbidden on windows)

**Comment by scottwittenburg** [comment]
> I think this PR is a good opportunity to discuss binary vs source.

Yes, makes sense, but it seems like potentially a big discussion.

> If you make those the same thing, you get the benefits of download cache that exists in "source mirrors" but does not in "binary mirrors".

Is the benefit you're referring to mostly that there's one code path for both things?  Or is it a benefit to have a local cache of downloaded tarballs you have fetched in the past?  That seems like it could get big, considering how many times you would expect to reuse any particular one.  For source tarballs, the benefit is clearer to me, since you might build many variants of the same version, and only have to fetch the source once.  Maybe you want to cache binary tarballs though, if you uninstall and re-install the same concrete spec a bunch of times?

> Also regarding naming scheme, follow either Spack's existing source cache structure, or a slight improvement on that, include the hash algorithm in the name too

The reason I haven't done that already is for ease of pruning by spec hash.  For me, the main point of this PR was to fix the race condition in CI, and do it in such that we can find and remove the orphaned tarballs without an exhaustive search.  But maybe you're saying the benefits of similar layout and fewer code paths outweigh any inconvenience of pruning this might introduce?

Since you've mentioned making source and binary mirrors the same, should we just focus on the source cache layout/naming (and let OCI naming be the odd one out)?  Or do you truly see either OCI or source layout/naming as equally good choices for these binaries?  Is there any drawback to having every known tarball in a single directory/prefix?  Can all three (source, binary, oci) use the same code and yet be laid out slightly differently?  Or is there already some ongoing effort or thought given to unifying the OCI vs source cache layouts?

> OCI scheme just uses e.g. sha256:<shasum> as a file name (except that the colon is probably forbidden on windows)

Doesn't OCI scheme use `sha256/<shasum>` for the naming?  In which case, there's no issue with `:` on windows.  But then, how would you want to see the layout?  Should it be laid out as `sha256/<2-char-prefix-of-sha256>/<sha256>` (and then source cache changed to match)?

Last question for now:  Do you want to resolve all these questions in this PR?

**Comment by scottwittenburg** [comment]
@tgamblin fyi, since we're discussing it on the call, regarding how the layout decision affects pruning

**Comment by haampie** [comment]
> Or is it a benefit to have a local cache of downloaded tarballs you have fetched in the past?

The current implementation of Gitlab CI is downloading worst case O(n^2) binaries to install O(n) packages from source ;) so caching binaries has use cases...

> But then, how would you want to see the layout? Should it be laid out as `sha256/<2-char-prefix-of-sha256>/<sha256>` (and then source cache changed to match)?

I checked git, which uses a fixed number of 256 subdirs `00 - FF` for object cache, which Spack probably copied. `ccache` uses dynamic depth.

The OCI directory layout does not have this extra subdir. I think they're focused on a single image (w/ multiple layer) per directory, instead of a full registry of multiple images in a directory. So maybe not the best to take inspiration from, also cause they [*require* an index.json file](https://github.com/opencontainers/image-spec/blob/main/image-layout.md#content) in this directory, which is a reduction over the cache contents, so blocking and annoying to deal with.

**Comment by kwryankrattiger** [comment]
I was thinking more about the layout with the meta files and I came up with a possible solution.

Introducing a "dag hash index file" that lists the content hash for a meta file and binary file pair. This file could be overwritten at any point, but the content pair that it referenced would always be consistent. This is similar to the much older build cache layout but decoupled. I think @haampie took a similar approach for embedding meta files in OCI.

```json
{
  "meta": {
    "hash": <sha256sum of meta file>
    "signed": true/false
  },
  "binary": {
    "hash": <sha256sum of binary file>
  }
}
```

Ex.

```console
index.json (optional)

index/
    jxosxtnpusbpptbk23mh66o6opxbahjt
    vly2lkyg5mnqxsyv4t24eme27mjbnwoz

sha256/
    4255d704de303a6a09c042dde7025ff539463c820a5111d85fb326cfed860570
    b2be6bee7c11375535258fcb18e00a616970a7e3efdb37312495e3bd92947634
```

* `66a045b452102c59d840ec097d59d9467e13a3f34f6494e539ffd32c1bb35f18`
  ```json
  {
    "meta": {
      "hash": "4255d704de303a6a09c042dde7025ff539463c820a5111d85fb326cfed860570"
    }
    "binary": {
      "hash": "b2be6bee7c11375535258fcb18e00a616970a7e3efdb37312495e3bd92947634"
    }
  }
  ```

* `4255d704de303a6a09c042dde7025ff539463c820a5111d85fb326cfed860570`
    ```
    Hello meta
    ```

* `b2be6bee7c11375535258fcb18e00a616970a7e3efdb37312495e3bd92947634`
  ```
  Hello binary
  ```

Update index and signing would have to iterate `index/*` to get the meta data content address, and then download the meta files. This should be only a relatively small amount of overhead in comparison to fetching the meta files directly by extension. The information from the `index/*` files could also probably go directly into the top level `index.json` as well.

Manual pruning would be much simpler, removing only the `index/<dag hash>` file would effectively make the associated content inaccessible after an `update-index`. This is much less error prone and time consuming compared to the current layout.

Automated pruning would now require downloading the files in `index/*`. However, the `direct` pruning method should be able to identify orphaned index and binary files directly rather than requiring a second approach that required downloading the entire `index.json` file.

Notice that the spec.json files no longer have any extension associated with them to denote "signed". This is intentional. The idea for that is two parts. a) it is in line with the content hash layout b) it is much easier to directly fetch the same object all the time and then check the first character of the contents to check if it is signed or not, in the case it is signed the character will be `-` with the header `-----BEGIN PGP SIGNED MESSAGE-----` and in the case it is not signed the first line is the beginning of the JSON dict `{`. 

Edit:

* Fixed DAG hash named files in checksum
* Add reasoning for dropping `.spec.json[.sig]` from all of the files

**Comment by scottwittenburg** [comment]
Thanks for your suggestion @kwryankrattiger :+1:

Some responses/thoughts: 

I agree the introduction of "dag hash index" is another way to avoid the race condition we currently have with concurrent pipeline jobs building the same dag hash.  And as you mentioned, it also seems like it would make manual and automated pruning a little easier, compared to an approach that names the metadata files themselves after the dag hash (and orphans only the binary instead of both the metadata and binary).

While it's true the new files are small, rebuilding the index would now require twice as many fetches.  Also, directly fetching a spec to install (still supported I think), requires three fetches instead of two in the simplest case.  Those details are likely not showstoppers, I'm just pointing them out.

> The information from the index/* files could also probably go directly into the top level index.json as well.

Maybe I misunderstand your point here, but I don't think this is the case.  We need to have the concrete spec json to add to the database instance that records remote specs, so I don't see how we can avoid fetching those files.

Additionally, I just want to be clear about a detail your code blocks above:  the sha-named files within the `index/` directory are not actually these new dag hash index files, right?  Instead they are what we currently store as the (possibly clear-signed) spec metadata files?  I would have expected the dag hash index files to be named using the dag hash, so they're either mis-named above, or they must live somewhere else.  Can you confirm that?  And maybe edit your post if necessary?

Overall, I think this approach could work.  I'm not sure whether it violates the advice from @haampie not to "invent a new naming scheme for each instance where something needs to be content addressable".  Also, it looks less like the source mirror layout/organization, which I thought the group expressed interest in, and more like the oci layout.  But actually, regarding the storage of binary blobs themselves, I guess it could be adjusted slightly to better match source mirror layout.

I think it would good to have some more discussion on your proposal here, and hopefully get some consensus on our proposed approach, whatever it is.  @tgamblin @alecbcs @zackgalbreath

**Comment by kwryankrattiger** [comment]
> Maybe I misunderstand your point here, but I don't think this is the case. We need to have the concrete spec json to add to the database instance that records remote specs, so I don't see how we can avoid fetching those files.

On that point, I am saying store the content hash in the top level `index.json` file directly. I don't know if that is actually useful or not. We certainly don't do this now for the checksum of the binary.

> Additionally, I just want to be clear about a detail your code blocks above: the sha-named files within the index/ directory are not actually these new dag hash index files, right?

There are two types of hashes here, the DAG hash (under `index/*`) and the content hash of what is being stored (under `sha256/*`).

In my example above I do show an example of the DAG hash index file which is named by the DAG hash under `index/66a045b452102c59d840ec097d59d9467e13a3f34f6494e539ffd32c1bb35f18`, it contains the content hash of a meta file and a binary file of which the contents are listed. The meta file is the object pointed to by the `meta:hash:4255d704de303a6a09c042dde7025ff539463c820a5111d85fb326cfed860570`. The contents of that object are not the actual format of the Spack meta files, I opted for simple text for brevity in the example.

> I'm not sure whether it violates the advice from @haampie not to "invent a new naming scheme for each instance where something needs to be content addressable".

I don't think this is inventing a new naming scheme for content addressed data. The prefix for the blob data in my example is not they key part of the proposal, it can be adjusted to match the current source cache layout (`<first two of content hash>/<content hash>`) or the OCI layout (`blobs/sha256/<content hash>`). The main thing here is both the meta data and the binary data are stored by content hash, and there is a "manifest" or "dag index file" that associates the content hash to the dag hash.

>  Also, it looks less like the source mirror layout/organization, which I thought the group expressed interest in, and more like the oci layout.

I think the main difference between the two is the OCI layout includes the hash algorithm used in the prefix. I think including the checksum algorithm in the name is a good thing to adopt for the build cache, and probably also for the source cache as well.

--

What I am trying to solve in this proposal is how to reconcile the two sources of hashes required for the build cache, the DAG hash and the data hash.

The motivation for a new file type is simply to make build cache pruning and inspection faster by reducing the size of the file that needs to be download to know which blobs are associated with a specified DAG hash. The main advantage of this method is pruning is slightly simpler than having to download, strip signatures, and parse the meta files. A secondary advantage could be that the "DAG index file" and the OCI manifest could be identical, which could more or less unify the code paths they use.

> While it's true the new files are small, rebuilding the index would now require twice as many fetches. Also, directly fetching a spec to install (still supported I think), requires three fetches instead of two in the simplest case. Those details are likely not showstoppers, I'm just pointing them out.

> > This should be only a relatively small amount of overhead in comparison to fetching the meta files directly by extension.

I agree that the extra OCI manifest-like DAG hash index file is extra overhead requiring additional fetching and parsing for each package. I am not entirely convinced that the extra complexity is worth the effort simply to match OCI or make pruning easier. While there are the advantages I listed above, indirection is complicated and can be confusing to follow.

**Comment by scottwittenburg** [comment]
> In my example above I do show an example of the DAG hash index file which is named by the DAG hash under index/66a045b452102c59d840ec097d59d9467e13a3f34f6494e539ffd32c1bb35f18

The things you have under `index/` aren't dag hashes, those are sha256 sums (dag hashes are 32 characters and use all the letters).  That was the source of my confusion.

> What I am trying to solve in this proposal is how to reconcile the two sources of hashes required for the build cache, the DAG hash and the data hash.

I see that your proposal accomplishes that.  I think it would also be accomplished by what we discussed (more or less) in our discussion a week ago:

```
build_cache/
    index.json (optional)

    jxosxtnpusbpptbk23mh66o6opxbahjt.spec.json[.sig]
    vly2lkyg5mnqxsyv4t24eme27mjbnwoz.spec.json[.sig]

    sha256/
        4255d704de303a6a09c042dde7025ff539463c820a5111d85fb326cfed860570
        b2be6bee7c11375535258fcb18e00a616970a7e3efdb37312495e3bd92947634
```

Where the `<dag_hash>.spec.json[.sig]` files contain the sha256 of the binary blob.  And to me it's not clear that the extra level of indirection you're proposing, which maybe makes pruning easier, is worth it.  Maybe it is.

Is there any difference in the arguments you're formulating elsewhere about *correctness* of pruning under either approach?

Edit: my suggestion here is focused on how we treat the metadata (the spec files), I don't have strong feelings one way or another about how the binary blobs are laid out.  So while my example above shows a binary layout similar to oci, it might make more sense to make the new binary layout reflect how source tarballs are currently stored.

**Comment by scottwittenburg** [comment]
closing in favor of #48713
